integration/before-after-request/simple.ts

```diff
@@ -0,0 +1,499 @@
+    message.kind = object.kind ?? 0;
+    return message;
+  },
+};
+
+messageTypeRegistry.set(FooServiceCreateRequest.$type, FooServiceCreateRequest);
+
+function createBaseFooServiceCreateResponse(): FooServiceCreateResponse {
+  return { $type: "simple.FooServiceCreateResponse", kind: 0 };
+}
+
+export const FooServiceCreateResponse = {
+  $type: "simple.FooServiceCreateResponse" as const,
+
+  encode(message: FooServiceCreateResponse, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
+    if (message.kind !== 0) {
+      writer.uint32(8).int32(message.kind);
+    }
+    return writer;
+  },
+
+  decode(input: _m0.Reader | Uint8Array, length?: number): FooServiceCreateResponse {
+    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
+    let end = length === undefined ? reader.len : reader.pos + length;
+    const message = createBaseFooServiceCreateResponse();
+    while (reader.pos < end) {
+      const tag = reader.uint32();
+      switch (tag >>> 3) {
+        case 1:
+          if (tag !== 8) {
+            break;
+          }
+
+          message.kind = reader.int32() as any;
+          continue;
+      }
+      if ((tag & 7) === 4 || tag === 0) {
+        break;
+      }
+      reader.skipType(tag & 7);
+    }
+    return message;
+  },
+
+  fromJSON(object: any): FooServiceCreateResponse {
+    return { $type: FooServiceCreateResponse.$type, kind: isSet(object.kind) ? fooServiceFromJSON(object.kind) : 0 };
+  },
+
+  toJSON(message: FooServiceCreateResponse): unknown {
+    const obj: any = {};
+    if (message.kind !== 0) {
+      obj.kind = fooServiceToJSON(message.kind);
+    }
+    return obj;
+  },
+
+  create<I extends Exact<DeepPartial<FooServiceCreateResponse>, I>>(base?: I): FooServiceCreateResponse {
+    return FooServiceCreateResponse.fromPartial(base ?? ({} as any));
+  },
+  fromPartial<I extends Exact<DeepPartial<FooServiceCreateResponse>, I>>(object: I): FooServiceCreateResponse {
+    const message = createBaseFooServiceCreateResponse();
+    message.kind = object.kind ?? 0;
+    return message;
+  },
+};
+
+messageTypeRegistry.set(FooServiceCreateResponse.$type, FooServiceCreateResponse);
+
+export interface FooService {
+  Create(request: FooServiceCreateRequest): Promise<FooServiceCreateResponse>;
+}
+
+export const FooServiceServiceName = "simple.FooService";
+export class FooServiceClientImpl implements FooService {
+  private readonly rpc: Rpc;
+  private readonly service: string;
+  constructor(rpc: Rpc, opts?: { service?: string }) {
+    this.service = opts?.service || FooServiceServiceName;
+    this.rpc = rpc;
+    this.Create = this.Create.bind(this);
+  }
+  Create(request: FooServiceCreateRequest): Promise<FooServiceCreateResponse> {
+    const data = FooServiceCreateRequest.encode(request).finish();
+    const requestMethod = messageTypeRegistry.get(request.$type);
+    if (this.rpc.beforeRequest && requestMethod) {
+      this.rpc.beforeRequest(requestMethod);
```

Comment:
I wasn't expecting to see the `messageTypeRegistry.get(request.$type)` lookup, b/c that just means the `beforeRequest` is going to be called the with metadata about the request, and not the actually `request` itself.

I.e. shouldn't this be:

```
this.rpc.beforeRequest(request);
```

And then if the caller wants to use `request.$type` to see what type of request it is, that's fine, but they also still have the request itself, to do any logging or manipulation of, as they'd like.

Reply:
Ah, yes. I think that's where I went wrong - I initially had `this.rpc.beforeRequest(request);` but got a bit confused between the `Request` function and `Request` interface, so when I was addressing the TS error I ended up moving towards the wrong functionality. Sending out an update with this fix shortly. 🙏🏼